lib/jxl/modular/transform/palette.h

```diff
@@ -443,23 +446,68 @@ static Status FwdPalette(Image &input, uint32_t begin_c, uint32_t end_c,
   std::vector<pixel_type> color(nb);
   std::vector<float> color_with_error(nb);
   std::vector<const pixel_type *> p_in(nb);
-  for (size_t y = 0; y < h; y++) {
-    for (uint32_t c = 0; c < nb; c++) {
-      p_in[c] = input.channel[begin_c + c].Row(y);
-    }
-    for (size_t x = 0; x < w; x++) {
-      if (lossy && candidate_palette.size() >= nb_colors) break;
+  if (nb > 1) {
+    for (size_t y = 0; y < h; y++) {
       for (uint32_t c = 0; c < nb; c++) {
-        color[c] = p_in[c][x];
+        p_in[c] = input.channel[begin_c + c].Row(y);
+      }
+      for (size_t x = 0; x < w; x++) {
+        if (lossy && candidate_palette.size() >= nb_colors) break;
+        for (uint32_t c = 0; c < nb; c++) {
+          color[c] = p_in[c][x];
+        }
+        const bool new_color = candidate_palette.insert(color).second;
+        if (new_color) {
+          candidate_palette_imageorder.push_back(color);
+        }
+        if (candidate_palette.size() > nb_colors) {
+          return false;  // too many colors
+        }
       }
-      const bool new_color = candidate_palette.insert(color).second;
-      if (new_color) {
-        candidate_palette_imageorder.push_back(color);
+    }
+  } else if (!lossy) {
```

Comment:
maybe we should move this up and make it an early return before we even create the std::set...

Reply:
Makes sense.

Reply:
Moved it up, PTAL.